notify/notify.go

```diff
@@ -119,109 +143,271 @@ func Now(ctx context.Context) (time.Time, bool) {
 	return v, ok
 }
 
-// A Notifier is a type which notifies about alerts under constraints of the
-// given context.
-type Notifier interface {
-	Notify(context.Context, ...*types.Alert) error
+// A Stage processes alerts constraints under the constraints of the given
+// context.
+type Stage interface {
+	Exec(ctx context.Context, alerts ...*types.Alert) ([]*types.Alert, error)
+}
+
+// Builds a RoutingStage as a map of receivers to Stages
+func BuildRoutingStage(confs []*config.Receiver, tmpl *template.Template, meshWait func() time.Duration, inhibitor *inhibit.Inhibitor, silences *meshprov.Silences, ni *meshprov.NotificationInfos, marker types.Marker) RoutingStage {
+	rs := make(RoutingStage)
+
+	for _, rc := range confs {
+		rs[rc.Name] = createStage(BuildReceiverIntegrations(rc, tmpl), meshWait, inhibitor, silences, ni, marker)
+	}
+
+	return rs
+}
+
+// createStage creates a MultiStage as a pipeline of stages for a receiver
+func createStage(receiverIntegrations []Integration, meshWait func() time.Duration, inhibitor *inhibit.Inhibitor, silences *meshprov.Silences, ni *meshprov.NotificationInfos, marker types.Marker) Stage {
+	var ms MultiStage
+	ms = append(ms, NewLogStage(log.With("step", "inhibit")))
+	ms = append(ms, NewInhibitStage(inhibitor, marker))
+	ms = append(ms, NewLogStage(log.With("step", "silence")))
+	ms = append(ms, NewSilenceStage(silences, marker))
+
+	var fs = make(FanoutStage)
+	for _, i := range receiverIntegrations {
+		var s MultiStage
+		s = append(s, NewLogStage(log.With("step", "wait")))
+		s = append(s, NewWaitStage(meshWait))
+		s = append(s, NewLogStage(log.With("step", "filterResolved")))
+		s = append(s, NewFilterResolvedStage(i.notifierConfig))
+		s = append(s, NewLogStage(log.With("step", "dedup")))
+		s = append(s, NewDedupStage(ni))
+		s = append(s, NewLogStage(log.With("step", "integration")))
+		s = append(s, NewRetryStage(i))
+		s = append(s, NewLogStage(log.With("step", "newNotifies")))
+		s = append(s, NewSetNotifiesStage(ni))
+		fs[fmt.Sprintf("%s/%d", i.name, i.idx)] = s
+	}
+
+	return append(ms, fs)
+}
+
+// RoutingStage executes the inner stages based on the receiver specified in
+// the context.
+type RoutingStage map[string]Stage
+
+// Exec implements the Stage interface.
+func (rs RoutingStage) Exec(ctx context.Context, alerts ...*types.Alert) ([]*types.Alert, error) {
+	receiver, ok := Receiver(ctx)
+	if !ok {
+		return nil, fmt.Errorf("receiver missing")
+	}
+
+	_, err := rs[receiver].Exec(ctx, alerts...)
```

Comment:
Panics if there is no entry for the key. This shouldn't happen by design and panics are fair to occur for programmer errors. 
But I think the coupling between what ensures this invariant and this line is too loose to trust in that. I'd add a check here. 

And a test :)